### ISSUE TITLE
request for git help

### ISSUE THREAD

**Issue by bdarcus** [citation-style-language/schema#215]
I've been merging commits for v1.1 in the v1.1 branch.

But I'm unsure of how to manage the git workflow so that, in the end, when we merge that branch to master, the commit history is correct.

In short, I want the new v1.1 commits to be added after the 1.0.x stuff, when they're ready, and to end up with correctly tagged v1.0.2 and v1.1 versions.

Related issue is a few in-progess PRs, where master changes during that progress.

You can also see an example of a PR that's otherwise ready to merge to master at #212.

My earlier assumption was that for the latter cases, the PR branches should merge (or rebase?) master.

Does this have anything to do with why the convention to do work on a `develop` branch; maybe so one can rebase on it to fix some of these issues?

@dhimmel - any advice?

References:

- https://git-scm.com/book/en/v2/Git-Branching-Rebasing
- https://spin.atomicobject.com/2018/10/10/git-branching-strategy/
- https://spin.atomicobject.com/2019/01/10/git-merging-vs-rebasing/

**Comment by dhimmel** [comment]
When you are ready to merge v1.1 branch into master, create a pull request at https://github.com/citation-style-language/schema/compare/v1.1. For this pull request, you will want to make a "merge commit" and not a squash commit like we recommended for other PRs. This is because you want to preserve the commit history on the v.1.1 branch

> In short, I want the new v1.1 commits to be added after the 1.0.x stuff

The v.1.1 branch commit will be added in parallel to the v1.0.x commits on master. Taken from a part of the graph at https://github.com/citation-style-language/schema/network

> ![image](https://user-images.githubusercontent.com/1117703/83354660-3a476c00-a328-11ea-8422-9678211650b7.png)

You see the two branches: black and blue. Black is master. You can have as many commits as you want on both branches. The merge commit brings the two branches together. GitHub is showing a combined order based on timestamps... but that's kind of an illusion, there are separate until merged.

I wouldn't worry about the commit order on GitHub following the merge. Does that make sense?

**Comment by bdarcus** [comment]
Got it on the merge commit. I figured that.

> I wouldn't worry about the commit order on GitHub following the merge. Does that make sense?

Two things.

1. so if I had a commit to the branch before a subsequent change on master, and then merge, that won't matter?
2. what do I do about #212, where csl.rnc is in conflict, presumably because master is ahead of it.

**Comment by dhimmel** [comment]
> so if I had a commit to the branch before a subsequent change on master, and then merge, that won't matter?

I don't quite understand the question, but I don't think there is anything you have to worry about.

> what do I do about #212, where csl.rnc is in conflict, presumably because master is ahead of it.

master having different commits alone is not sufficient to cause conflicts. Both branches have changes to the same lines of code. I am surprised that you have a conflict on #212. Usually moving a file is okay. One way you could approach this is to merge `master` into ` restructure-merge` to fix the conflict. Then I can take a look to see if it looks fixed, before you ultimately squash merge #212 removing the messy merge commits

**Comment by bdarcus** [comment]
Merging from master indeed fixed it.

So for PRs that take time to develop, authors should in fact merge from master if they run into conflicts, and we can just squash the history merge to master.

Right?

**Comment by dhimmel** [comment]
> So for PRs that take time to develop, authors should in fact merge from master if they run into conflicts, and we can just squash the history when we merge to master.

Yes.

Unless the merge is for something like `v1.1` where you want to keep commits separate. It's also possible to rebase a branch on master, but a bit more risky if you mess up.

> So for PRs that take time to develop, authors should in fact merge from master if they run into conflicts

Note that conflicts are actually pretty rare unless changes are large and sweeping or PRs are kept open for long periods of time. You're getting more now because of the reorganization, lot's of whitespace changes, etcetera.

In general, it's the onus of the PR creator to resolve conflicts, but it's best for maintainers to be cognizant of what actions will create unnecessary conflicts.

**Comment by bdarcus** [comment]
Another (no rush) question related to the v1.1 branch.

If I want to revise some of the history of commits specific to that branch (new features and such) to ensure greater consistency, and tie in with changelog management (#220) going forward, can I use rebase on that branch?

Or maybe when we get closer to release (probably later in the Summer), I can ask you to review and help get the git details correct?

**Comment by dhimmel** [comment]
You could `git rebase --interactive master`. Make sure to duplicate the branch before doing this in case you mess things up. It can be confusing and error prone... but if you do things right, anything is possible.

There are many drawbacks to rewriting history. Lot's of comments or links on GitHub that refer to specific commit hashes will get messed up. I would avoid rewriting on master except in extreme circumstances. The longer content has been on a branch, the worse it is to rewrite.

As far as rewriting v1.1, I personally would focus most on making sure future commits follow all the conventions you want, rather than rewriting past ones from already merged PRs. Although you'll be able to judge this better than me.

**Comment by bdarcus** [comment]
> As far as rewriting v1.1, I personally would focus most on making sure future commits follow all the conventions you want, rather than rewriting past ones from already merged PRs.

OK, so going forward, is it wise to suggest the following?

- PRs should be focused, so we can typically do squash commits, where the commit messages are clear
- for a release branch (like 1.1) we should plan to do a standard merge commit
- rebase commits are problematic enough that we should turn them off on the repo?

**Comment by denismaier** [comment]
Will er really need to keep updating 1.0 once 1.1 is out? I thought 1.1 will be merged into master once it is ready, tagged 1.1.0 and then development takes place there. Any non minor PRs go then in a 1.2 branch.
